### ISSUE TITLE
MSVC static runtime library

### ISSUE THREAD

**Issue by jsutes** [noloader/cryptopp-cmake#63]
According to https://www.cryptopp.com/wiki/Visual_Studio cryptopp should favor static linking against the C/C++ runtime (/MT or /MTd).

When I use this cmake project with visual studio or ninja generator, I keep getting dynamically linked runtime. You can test this by searching the build.ninja output file for "\MD" or "\MT".

My cmake command is

    cmake -G Ninja -DBUILD_SHARED=OFF -DBUILD_STATIC=ON -DBUILD_TESTING=OFF -DCMAKE_INSTALL_PREFIX=C:\cryptopp -DCMAKE_BUILD_TYPE=Release ..

How can we switch between static and dynamic linking with this project?
According to https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html#variable:CMAKE_MSVC_RUNTIME_LIBRARY and https://cmake.org/cmake/help/latest/prop_tgt/MSVC_RUNTIME_LIBRARY.html
we should be able to use ``CMAKE_MSVC_RUNTIME_LIBRARY`` as a global variable or ``MSVC_RUNTIME_LIBRARY`` as a target specific property to choose the desired option.  I tried setting these variables to "MultiThreaded" but it did not have an effect (I do successfully use these variables in my own projects).

I was able to do a find-and-replace "/MD" to "/MT" in the build.ninja file and successfully built with static linked runtime library, so there shouldn't be issues with the build. We just need to figure out how to switch between the two in cmake instead of having to manually edit the ninja or visual studio output file.

OS: Windows 10
Compiler: MSVC/14.28.29333
CMake: 3.19

**Comment by CleanHit** [comment]
@jsutes Here is my _CMakeFiles.txt_. With it I download and build the `cryptopp` static library using **Visual Studio Community 2019**s _MSCV_ compiler in **Visual Studio Code** in **Windows**, using **MinGWs** x64 _GCC_ compiler in **Windows** was just a pain in the $%&.
```cmake
cmake_minimum_required(VERSION 3.15 FATAL_ERROR)
project(problem_child CXX C)

# Build in Debug mode by default
if(NOT CMAKE_BUILD_TYPE)
    set(CMAKE_BUILD_TYPE Debug)
endif()
message(STATUS "Build type is set to: ${CMAKE_BUILD_TYPE}")

# Set compiler flags
set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ standard to be used")
if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang") # Using Clang compiler
    message(STATUS "Using Clang compiler")
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O1 -g -D_DEBUG -Wall -Wextra -Wconversion -fopenmp")
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -g")
elseif (CMAKE_CXX_COMPILER_ID STREQUAL "GNU") # Using GCC compiler
    message(STATUS "Using GNU compiler")
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O1 -g -D_DEBUG -Wall -Wextra -Wconversion -fopenmp")
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -g")
elseif (CMAKE_CXX_COMPILER_ID STREQUAL "Intel") # Using Intel C++ compiler
    message(STATUS "Using Intel compiler")
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O1 -D_DEBUG -traceback -xarch -g -debug all")
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -g -xarch")
elseif (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC" AND CMAKE_HOST_WIN32) # Using Visual Studio C++ compiler
    message(STATUS "Using MSVC compiler")
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /Od /EHsc /Zi /GS- /Zc:__cplusplus /W4 /D_DEBUG")
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /O2 /Z7")
else()
    message(WARNING "A diffrent compiler than Clang, GNU, Intel or MSVC (on Windows) was detected.")
endif()

# Include usage of external projects
include(FetchContent)

############################################################################################
# Get the CryptoPP CMakeLists.txt File for CryptoPP Package GIT_TAG Must be the Same
############################################################################################
message(CHECK_START "Fetching CryptoPP-CMAKE")
set(CRYPTOPP_CMAKE "cryptopp-cmake")
set(CRYPTOPP_CMAKE_GIT_TAG "master") # CRYPTOPP_8_2_0
set(CRYPTOPP_GIT_TAG "CRYPTOPP_8_3_0")
FetchContent_Declare(
        ${CRYPTOPP_CMAKE}
        GIT_REPOSITORY  https://github.com/noloader/cryptopp-cmake
        GIT_TAG         ${CRYPTOPP_CMAKE_GIT_TAG}
)

FetchContent_GetProperties(${CRYPTOPP_CMAKE})
if(NOT ${CRYPTOPP_CMAKE}_POPULATED)
    FetchContent_Populate(${CRYPTOPP_CMAKE})
endif()

############################################################################################
# Get the CryptoPP Package
############################################################################################
message(CHECK_START "Fetching CryptoPP")
set(CRYPTOPP "cryptopp")
FetchContent_Declare(
        ${CRYPTOPP}
        GIT_REPOSITORY  https://github.com/weidai11/cryptopp
        GIT_TAG         ${CRYPTOPP_GIT_TAG}
)

FetchContent_GetProperties(${CRYPTOPP})
if(NOT ${CRYPTOPP}_POPULATED)
    FetchContent_Populate(${CRYPTOPP})
    # !!! IMPORTANT !!!
    # Copy the CMakeLists.txt file from https://github.com/noloader/cryptopp-cmake with same TAG CRYPTOPP repository into ${${CRYPTOPP}_SOURCE_DIR}
    # before using add_subdirectory() and include_directories() commands, until the a proper COPY command was implemented.
    file(COPY ${${CRYPTOPP_CMAKE}_SOURCE_DIR}/CMakeLists.txt DESTINATION ${${CRYPTOPP}_SOURCE_DIR})
    add_subdirectory(${${CRYPTOPP}_SOURCE_DIR} ${${CRYPTOPP}_BINARY_DIR})
endif()
include_directories(${${CRYPTOPP}_SOURCE_DIR})

# Set Cryptopp library properties every time after the first population
if(${CRYPTOPP}_POPULATED)
    # Build shared or static library
    set(BUILD_SHARED_CRYPTOPP_OLD ${BUILD_SHARED})
    set(BUILD_SHARED OFF CACHE INTERNAL "Build CryptoPP SHARED libraries")
    message(STATUS "Build CryptoPP shared lib is set to: ${BUILD_SHARED}")
    if(${BUILD_SHARED} STREQUAL "ON")
        set(CRYPTOPP "cryptopp-shared")
    else()
        set(CRYPTOPP "cryptopp-static")
    endif()
    set(BUILD_SHARED ${BUILD_SHARED_CRYPTOPP_OLD} CACHE BOOL "Type of libraries to build" FORCE)
endif()

# Executable will be in ../bin
message(CHECK_START "Create an executable problem_child")
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG ${PROJECT_SOURCE_DIR}/bin/debug)
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELEASE ${PROJECT_SOURCE_DIR}/bin/release)

set(APP_DOMAIN ${PROJECT_SOURCE_DIR})
add_executable(problem_child ${APP_DOMAIN}/main.cpp)

# Link the external libraries to the executable
message(CHECK_START "Linking all dependency libraries to problem_child")
target_link_libraries(problem_child PRIVATE 
	${CRYPTOPP} # CryptoPP static or shared lib
)
```
Unfortunately, I can't build the `cryptopp` shared library in **Windows**, in **Linux** everything works fine with _GCC_. When I change
```cmake
set(BUILD_SHARED OFF CACHE INTERNAL "Build CryptoPP SHARED libraries")
```
To
```cmake
set(BUILD_SHARED ON CACHE INTERNAL "Build CryptoPP SHARED libraries")
```
I get this build error
```bash
[main] Building folder: problem-child 
[build] Starting build
[proc] Executing command: C:\msys64\mingw64\bin\cmake.exe --build f:/problem-child/build --config Debug --target problem_child -- -j 6
[build] [ 96%] Built target cryptopp-object
[build] [ 98%] Built target cryptopp-shared
[build] mingw32-make[3]: *** No rule to make target '_deps/cryptopp-build/cryptopp-shared.lib', needed by '../bin/debug/problem_child.exe'.  Stop.
[build] mingw32-make[2]: *** [CMakeFiles\Makefile2:124: CMakeFiles/problem_child.dir/all] Error 2
[build] mingw32-make[1]: *** [CMakeFiles\Makefile2:131: CMakeFiles/problem_child.dir/rule] Error 2
[build] mingw32-make: *** [Makefile:182: problem_child] Error 2
[build] Build finished with exit code 2
```
Does anyone know what _cryptopp-cmake_ **CMake** variable I should set to build the shared library?

My environment info:
**OS:** Windows 10
**Compiler:** MSVC/14.28.29333
**CMake:** 3.19.1

**Comment by noloader** [comment]
> using MinGWs x64 GCC compiler in Windows was just a pain in the $%&.

CMake is a royal pain in the $%&.

> Here is my CMakeFiles.txt.

You can zip `CMakeFiles.txt` and then drag-and-drop the zip file in the bug report to make it easy to download.

**Comment by CleanHit** [comment]
> > using MinGWs x64 GCC compiler in Windows was just a pain in the $%&.
> 
> CMake is a royal pain in the $%&.
> 
> > Here is my CMakeFiles.txt.
> 
> You can zip `CMakeFiles.txt` and then drag-and-drop the zip file in the bug report to make it easy to download.

You are correct and I've updated my post.

**Comment by jsutes** [comment]
Thanks for the comments.
I've continued building cryptopp statically with dynamic linking against the C/C++ runtime even though it is against the advice of https://www.cryptopp.com/wiki/Visual_Studio. It's actually made my builds easier since other third party libraries I use get built with dynamic C/C++ runtime by default.

@CleanHit

> I get this build error build_cryptopp_shared_error.txt Does anyone know what cryptopp-cmake CMake variable I should set to build the shared library?

So far when I want to use cryptopp as a shared library, I've been using vcpkg. You say your compiler is MSVC but your error txt file looks like your using mingw32-make as the build tool. Is that allowed? I've only used Ninja or Visual Studio as the build tool with MSVC and CMake.

In summary, I'm fine if this issue is closed since it's no longer a problem for me. But maybe it should be kept open until an option is added to toggle static/dynamic C/C++ runtime linking.

**Comment by CleanHit** [comment]
> So far when I want to use cryptopp as a shared library, I've been using vcpkg. You say your compiler is MSVC but your error txt file looks like your using mingw32-make as the build tool. Is that allowed? I've only used Ninja or Visual Studio as the build tool with MSVC and CMake.

Thanks for point that out. It was the remaining of my attempts to compile using _MSYS2_s **GNU** compiler with `mingw-w64-x86_64-cmake`, `mingw-w64-x86_64-make`, `mingw-w64-x86_64-gcc`, `mingw-w64-x86_64-gdb` and `mingw-w64-x86_64-clang` packages. The only thing that was left from that setting was `mingw-w64-x86_64-cmake` and `mingw-w64-x86_64-make` packages. I've removed that mow and set the  _CMake_ standalone installation for **Windows** to be used.

> I've only used Ninja or Visual Studio as the build tool with MSVC and CMake.

I couldn't make it compile using **Ninja** in **Visual Studio Community 2019** and **Visual Studio Code**, I was getting the error `...CMakeFiles\rules.ninja:293: expected '=', got lexing error out\build\x64-Release...`. I had to set my _CMake_ generator to `"generator": "Visual Studio 16 2019 Win64"` in **Visual Studio Community 2019** or to `"cmake.generator": "Visual Studio 16 2019"` in global **VSCode** user _settings.json_ file.

If I try to build `cryptopp-shared` lib in **Visual Studio Community 2019** the file is not being built and I get:
```bash
Severity Code Description Project File Line Suppression State Error LNK1104 cannot open file '_deps\cryptopp-build\Debug\cryptopp-shared.lib'  <path_to_project>\out\build\x64-Debug (Standard)\hmmenc_client <path_to_project>\out\build\x64 Debug (Standard)\LINK 1
```
Only _cryptopp-shared.dll_, _cryptopp-shared.ilk_ and _cryptopp-shared_ get created in *<path_to_project>\out\build\x64-Debug (Standard)\_deps\cryptopp-build\Debug*

**Comment by noloader** [comment]
Crypto++ DLL is probably not going to work. It is actually a [FIPS DLL](https://www.cryptopp.com/wiki/FIPS_DLL) and it is a lot of trouble. My advice is, remove anything that has to do with a DLL.

If you want a DLL, then create a [Wrapper DLL](https://www.cryptopp.com/wiki/Wrapper_DLL) with the functions you want to use.

With that said, I don't know what the CMake folks are doing. I keep this repository open so there's a central location for CMakeFileList.txt. People can callborate on it, and folks know the one place to download it.
